subprojects/core/src/main/java/org/gradle/internal/execution/fingerprint/impl/DefaultFileCollectionFingerprinterRegistry.java

```diff
@@ -17,29 +17,61 @@
 package org.gradle.internal.execution.fingerprint.impl;
 
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
+import org.gradle.api.tasks.ClasspathNormalizer;
+import org.gradle.api.tasks.CompileClasspathNormalizer;
+import org.gradle.api.tasks.FileNormalizer;
 import org.gradle.internal.execution.fingerprint.FileCollectionFingerprinter;
 import org.gradle.internal.execution.fingerprint.FileCollectionFingerprinterRegistry;
 import org.gradle.internal.execution.fingerprint.FileNormalizationSpec;
+import org.gradle.internal.fingerprint.DirectorySensitivity;
+import org.gradle.internal.fingerprint.IgnoredPathInputNormalizer;
+import org.gradle.internal.fingerprint.LineEndingNormalization;
 
 import java.util.Collection;
 import java.util.Map;
+import java.util.Set;
 
 import static org.gradle.internal.execution.fingerprint.impl.DefaultFileNormalizationSpec.from;
 
 public class DefaultFileCollectionFingerprinterRegistry implements FileCollectionFingerprinterRegistry {
     private final Map<FileNormalizationSpec, FileCollectionFingerprinter> fingerprinters;
 
+    /**
+     * These are the normalizers where it does not make sense to ignore empty directories.  If the user specifies non-sensical empty directory
+     * normalization (e.g. via the runtime api) we will ignore that criteria when selecting a fingerprinter.
+     */
+    private final static Set<Class<? extends FileNormalizer>> DIRECTORY_INSENSITIVE_NORMALIZERS = ImmutableSet.of(
+        ClasspathNormalizer.class,
+        CompileClasspathNormalizer.class,
+        IgnoredPathInputNormalizer.class
+    );
+
+    /**
+     * These are the normalizers where it does not make sense to ignore line endings.  If the user specifies non-sensical line ending
+     * normalization (e.g. via the runtime api) we will ignore that criteria when selecting a fingerprinter.
+     */
```

Comment:
It feels like we should warn the user in some way. With this newly introduced annotation failing the build would be okay I think.

Reply:
I think this came up with directory sensitivity, too.  I feel a little uncomfortable throwing a warning here because what they are asking for is not "wrong", it's just redundant.  I can ask for the default behavior (by specifying nothing) or I can ask explicitly for line endings to be ignored (by calling `.ignoreLineEndings()`).  Either way, I'm asking for the same thing when combined with these normalizers.

I could be convinced that a warning is better if you feel strongly about this.

Reply:
If it is all the same, then why do we need special processing here? Can't we just enable line ending normalization, and be done with it?

Reply:
I don't understand this comment.  We have fingerprinters that care about line ending normalization in text files and then we have one (CompileClasspathNormalizer) where it doesn't care because we're always looking at binary class files.  This is just handling the case where the user specifies criteria that says "I want compile classpath normalization and I want to ignore line endings in text files".  In that case, the default behavior and the behavior that has been asked for is the same behavior.

One option for dealing with this is to do what we are currently doing (i.e. just normalize the criteria in the registry).

Another option is that we could just register the same fingerprinter multiple times with different criteria, but that would just be moving the problem somewhere else (i.e. we'd still have to keep track of which fingerprinters need to be registered this way).

A third option would be to register multiple fingerprinters with different criteria for each.  This means we don't need any special processing, but when you combine this with the same approach for directory sensitivity, you end up with a bunch of different fingperinter objects in the registry that all do the same thing, but just have different criteria.  

Or you can throw an error, which is a little wonky because it's not really wrong, it's just over-precise.

Seems like an array of poor choices to me and I picked the one that felt the "least bad".  If you feel strongly that this should be an error, I'm fine with going that route, too.  Or if there is another option I'm not thinking of, I'm open to suggestions.

Reply:
Ok, I've refactored this so that we register a fingerprinter for each combination of criteria but we register the same fingerprinter in the case where it is not sensitive to a criterion.  e.g. There is only one `CompileClasspathFingerprinter` for all combinations of directory and line ending sensitivity and there is only one `IgnoredPathFingerprinter` for each line ending sensitivity value which gets registered for both directory sensitivity values, etc.